### ISSUE TITLE
Separate rule for non-required TypedDict field access

### ISSUE THREAD

**Issue by tekumara** [microsoft/pyright#1693]
pyright 1.1.120 introduced a check on field access for TypedDicts with `total=false` (see https://github.com/microsoft/pyright/issues/1584).

The [boto stubs ](https://github.com/vemel/mypy_boto3_builder) use `total=false` for TypedDicts, and so the following code:
```python
from typing import List
import boto3

s3_client = boto3.client("s3")

response = s3_client.list_buckets()

bucket_names: List[str] = [b["Name"] for b in response["Buckets"]]
```

Produces:
```
test.py:8:47 - error: Could not access item in TypedDict
    "Buckets" is not a required key in "ListBucketsOutputTypeDef", so access may result in runtime exception (reportGeneralTypeIssues)
test.py:8:28 - error: Could not access item in TypedDict
    "Name" is not a required key in "BucketTypeDef", so access may result in runtime exception (reportGeneralTypeIssues)
```

Would it be possible to separate this check into its own configurable rule?

**Comment by tekumara** [comment]
FYI - looks like an [issue has been raised on the boto stubs ](https://github.com/vemel/mypy_boto3_builder/issues/70) to adopt PEP 655.

**Comment by erictraut** [comment]
PEP 655 is not required to define "required" and "not required" fields in a TypedDict. You can use type inheritance to solve this. It appears that the generated boto3 stubs do already use this technique in some cases. For example:

```python
_RequiredCORSRuleTypeDef = TypedDict(
    "_RequiredCORSRuleTypeDef",
    {"AllowedMethods": List[str], "AllowedOrigins": List[str]},
)
_OptionalCORSRuleTypeDef = TypedDict(
    "_OptionalCORSRuleTypeDef",
    {
        "ID": str,
        "AllowedHeaders": List[str],
        "ExposeHeaders": List[str],
        "MaxAgeSeconds": int,
    },
    total=False,
)

class CORSRuleTypeDef(_RequiredCORSRuleTypeDef, _OptionalCORSRuleTypeDef):
    pass
```

It looks like this technique needs to be extended to other types like the ones used in your code sample above.

That said, I anticipated that we might receive a request to move this particular check to its own configurable rule. My hope was that TypedDict was sufficiently new and well-documented that there was little or no incorrect usage of it in shipping stubs. It appears that's not the case. Therefore, I think it's reasonable to split this diagnostic check to a separate rule.

**Comment by erictraut** [comment]
The name of the new diagnostic rule is "reportTypedDictNotRequiredAccess", and its default severity is "error". Now that it's split out from "reportGeneralTypeIssue", you can control it independently. This support will be in the next release of pyright and pylance.

**Comment by erictraut** [comment]
This is included in pyright 1.1.127, which I just published. It will also be included in the next release of pylance.
